src/style/properties.js

```diff
@@ -445,8 +446,8 @@ export class PossiblyEvaluatedPropertyValue<T> {
         }
     }
 
-    evaluate(feature: Feature): T {
-        return this.property.evaluate(this.value, this.globals, feature);
+    evaluate(feature: Feature, featureState?: FeatureState): T {
```

Comment:
Should `featureState` be a required parameter everywhere that `feature` is required?

Reply:
The `feature-state` expression is only usable in data-driven paint properties, so I restricted the change to that code path. The one place where this might be useful is  `Binder<T>#populatePaintArray`. But that method is run on the worker, and we don't have state available there, so it would also be unused.

Reply:
The benefit of making it required is that it leans on the type system to enforce consideration of of whether state needs to be provided.

* We can be sure we've looked at all the existing code paths.
* Code paths where state isn't available make that explicit by passing an empty object.
* If we add another evaluation code path, we can be sure we don't forget state where it should be supplied.